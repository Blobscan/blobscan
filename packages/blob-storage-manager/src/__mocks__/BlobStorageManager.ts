import { beforeEach } from "vitest";
import { mockDeep, mockReset } from "vitest-mock-extended";

import type { BlobStorageManager } from "../BlobStorageManager";

beforeEach(() => {
  mockReset(blobStorageManager);
});

const blobStorageManager = mockDeep<BlobStorageManager>();
export default blobStorageManager;
