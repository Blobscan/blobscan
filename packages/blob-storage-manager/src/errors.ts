class ErrorException extends Error {
  constructor(message: string, cause?: Error | Error[]) {
    super(message, {
      cause,
    });

    this.name = this.constructor.name;
  }
}

export class BlobStorageError extends ErrorException {
  constructor(storageName: string, message: string, cause?: Error) {
    super(`${storageName} failed: ${message}`, cause);
  }
}

export class StorageCreationError extends BlobStorageError {
  constructor(storageName: string, message: string, cause: Error) {
    super(
      storageName,
      `Creation failed${message.length ? `: ${message}` : ""}`,
      cause
    );
  }
}

export class BlobStorageManagerError extends ErrorException {
  constructor(message: string, cause?: Error | BlobStorageError[]) {
    super(message, cause);
  }
}
