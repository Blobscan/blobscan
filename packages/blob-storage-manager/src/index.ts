export { BlobStorageManager } from "./BlobStorageManager";
export * from "./blob-storage-manager";
export { env } from "./env";
export type { Environment } from "./env";
export * from "./errors";
export { GoogleStorage, PostgresStorage, SwarmStorage } from "./storages";
export * from "./types";
