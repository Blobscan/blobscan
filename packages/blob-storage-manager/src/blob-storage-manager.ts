import { logger } from "@blobscan/logger";

import type { BlobStorage } from "./BlobStorage";
import { BlobStorageManager } from "./BlobStorageManager";
import { env } from "./env";
import type { Environment } from "./env";
import type { BlobStorageName } from "./types";
import { BLOB_STORAGE_NAMES, createStorageFromEnv } from "./utils";

let blobStorageManager: BlobStorageManager | undefined;

function isBlobStorageEnabled(storageName: BlobStorageName) {
  const storageEnabledKey =
    `${storageName}_STORAGE_ENABLED` as keyof Environment;
  const storageEnabled = env[storageEnabledKey];

  return storageEnabled === true || storageEnabled === "true";
}

async function createBlobStorageManager() {
  const blobStorages = await Promise.all(
    Object.values(BLOB_STORAGE_NAMES).map(async (storageName) => {
      if (isBlobStorageEnabled(storageName)) {
        const [storage, storageError] = await createStorageFromEnv(storageName);

        if (storageError) {
          logger.warn(
            `${storageError.message}. Caused by: ${storageError.cause}`
          );
        }

        return storage;
      }
    })
  );

  const availableStorages = blobStorages.filter(
    (storage): storage is BlobStorage => !!storage
  );

  return new BlobStorageManager(availableStorages);
}

export async function getBlobStorageManager(): Promise<BlobStorageManager> {
  if (!blobStorageManager) {
    blobStorageManager = await createBlobStorageManager();
  }

  return blobStorageManager;
}
