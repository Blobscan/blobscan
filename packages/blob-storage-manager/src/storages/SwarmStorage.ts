import { Bee, BeeDebug } from "@ethersphere/bee-js";
import type { PostageBatch } from "@ethersphere/bee-js";

import type { BlobStorageConfig } from "../BlobStorage";
import { BlobStorage } from "../BlobStorage";
import type { Environment } from "../env";
import { BlobStorageError } from "../errors";
import { BLOB_STORAGE_NAMES } from "../utils";

export interface SwarmStorageConfig extends BlobStorageConfig {
  beeEndpoint: string;
  beeDebugEndpoint?: string;
}

export type SwarmClient = {
  bee: Bee;
  beeDebug?: BeeDebug;
};

export class SwarmStorage extends BlobStorage {
  _swarmClient: SwarmClient;

  constructor({ beeDebugEndpoint, beeEndpoint, chainId }: SwarmStorageConfig) {
    super(BLOB_STORAGE_NAMES.SWARM, chainId);

    try {
      this._swarmClient = {
        bee: new Bee(beeEndpoint),
        beeDebug: beeDebugEndpoint ? new BeeDebug(beeDebugEndpoint) : undefined,
      };
    } catch (err) {
      throw new Error("Failed to create swarm clients", {
        cause: err,
      });
    }
  }

  async getPostageBatch(batchLabel: string): Promise<PostageBatch | undefined> {
    const batches = await this.#getAllPostageBatch();

    return batches.find((b) => b.label === batchLabel);
  }

  protected async _healthCheck() {
    const healthCheckOps = [];

    healthCheckOps.push(this._swarmClient.bee.checkConnection());

    if (this._swarmClient.beeDebug) {
      healthCheckOps.push(
        this._swarmClient.beeDebug.getHealth().then((health) => {
          if (health.status !== "ok") {
            throw new Error(`Bee debug is not healthy: ${health.status}`);
          }
        })
      );
    }

    await Promise.all(healthCheckOps);
  }

  protected async _getBlob(reference: string) {
    const file = await this._swarmClient.bee.downloadFile(reference);

    return file.data.text();
  }

  protected async _storeBlob(versionedHash: string, data: string) {
    const batchId = await this.#getAvailableBatch();
    const response = await this._swarmClient.bee.uploadFile(
      batchId,
      data,
      this.buildBlobFileName(versionedHash),
      {
        pin: true,
        contentType: "text/plain",
      }
    );

    return response.reference.toString();
  }

  async #getAllPostageBatch(): Promise<PostageBatch[]> {
    if (!this._swarmClient.beeDebug) {
      throw new Error("Bee debug endpoint required to get postage batches.");
    }

    return this._swarmClient.beeDebug.getAllPostageBatch();
  }

  async #getAvailableBatch(): Promise<string> {
    const [firstBatch] = await this.#getAllPostageBatch();

    if (!firstBatch?.batchID) {
      throw new Error("No postage batches available.");
    }

    return firstBatch.batchID;
  }

  static getConfigFromEnv(env: Partial<Environment>): SwarmStorageConfig {
    const baseConfig = super.getConfigFromEnv(env);

    if (!env.BEE_ENDPOINT) {
      throw new BlobStorageError(
        this.name,
        "No endpoint provided. You need to define BEE_ENDPOINT variable in order to use the Swarm storage"
      );
    }

    return {
      ...baseConfig,
      beeDebugEndpoint: env.BEE_DEBUG_ENDPOINT,
      beeEndpoint: env.BEE_ENDPOINT,
    };
  }
}
