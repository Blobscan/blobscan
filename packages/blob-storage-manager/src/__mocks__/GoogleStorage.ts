import type { Storage } from "@google-cloud/storage";
import type { StorageOptions } from "@google-cloud/storage";
import { vi } from "vitest";

import { FILE_URI } from "../../test/fixtures";
import { GoogleStorage } from "../storages";
import type {
  GoogleCredentials,
  GoogleStorageConfig,
} from "../storages/GoogleStorage";

vi.mock("@google-cloud/storage", async () => {
  return {
    Storage: vi.fn().mockImplementation((config: StorageOptions) => {
      return {
        projectId: config.projectId,
        apiEndpoint: config.apiEndpoint,
        getServiceAccount: vi.fn().mockReturnValue([
          {
            emailAddress: config.credentials
              ? (config.credentials as GoogleCredentials).client_email
              : undefined,
          },
        ]),
        getBuckets: vi.fn().mockResolvedValue([[{ name: "mock-bucket" }]]),
        bucket: vi.fn().mockReturnValue({
          file: vi.fn().mockImplementation((uri: string) => {
            return {
              download: vi.fn().mockImplementation(() => {
                if (uri === FILE_URI) {
                  return Buffer.from("mock-data");
                }
                throw new Error("File not found");
              }),
              save: vi.fn((data: string) => data),
            };
          }),
        }),
        createBucket: vi.fn().mockResolvedValue(undefined),
      };
    }),
  };
});

export class GoogleStorageMock extends GoogleStorage {
  constructor(config: GoogleStorageConfig) {
    super(config);
  }

  get bucketName(): string {
    return this._bucketName;
  }

  get storageClient(): Storage {
    return this._storageClient;
  }
}
