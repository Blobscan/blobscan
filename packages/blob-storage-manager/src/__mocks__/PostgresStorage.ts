import type { PrismaClient } from "@blobscan/db";
import prisma from "@blobscan/db/prisma/__mocks__/client";

import { BLOB_HASH, RAW_DATA } from "../../test/fixtures";
import { PostgresStorage } from "../storages";

export class PostgresStorageMock extends PostgresStorage {
  constructor() {
    super();

    this.client = prisma as unknown as PrismaClient;
  }

  getBlob(versionedHash: string): Promise<string> {
    // mock prisma query
    prisma.blobData.findFirstOrThrow.mockResolvedValue({
      id: BLOB_HASH,
      data: RAW_DATA,
    });

    return super.getBlob(versionedHash);
  }

  async storeBlob(
    _: number,
    versionedHash: string,
    blobData: string
  ): Promise<string> {
    // mock prisma query
    prisma.blobData.upsert.mockResolvedValue({
      id: BLOB_HASH,
      data: RAW_DATA,
    });

    return super.storeBlob(_, versionedHash, blobData);
  }
}
