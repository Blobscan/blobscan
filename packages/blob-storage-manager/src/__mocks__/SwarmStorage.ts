import { Bee, BeeDebug } from "@ethersphere/bee-js";
import { vi } from "vitest";

import { SWARM_REFERENCE } from "../../test/fixtures";
import type { SwarmClient, SwarmStorageConfig } from "../storages";
import { SwarmStorage } from "../storages";

vi.mock("@ethersphere/bee-js", async () => {
  return {
    Bee: vi.fn().mockImplementation((endpoint) => {
      return {
        url: endpoint,
        checkConnection: vi
          .fn()
          .mockResolvedValueOnce({
            status: "ok",
          })
          .mockRejectedValueOnce(new Error("Bee is not healthy: not ok"))
          .mockResolvedValueOnce({
            status: "ok",
          }),
        downloadFile: vi.fn().mockImplementation((reference) => {
          if (reference === SWARM_REFERENCE) {
            return {
              data: {
                text() {
                  return "mock-data";
                },
              },
            };
          }
          throw new Error("File not found");
        }),
        uploadFile: vi.fn().mockResolvedValue({
          reference: "mock-reference",
        }),
      };
    }),
    BeeDebug: vi.fn().mockImplementation((endpoint) => {
      return {
        url: endpoint,
        getHealth: vi
          .fn()
          .mockResolvedValueOnce({
            status: "ok",
          })
          .mockResolvedValueOnce({
            status: "ok",
          })
          .mockResolvedValueOnce({
            status: "not ok",
          }),
        getAllPostageBatch: vi
          .fn()
          .mockResolvedValueOnce([
            {
              batchID: "mock-batch-id",
            },
          ])
          .mockResolvedValueOnce([])
          .mockResolvedValueOnce([]),
      };
    }),
  };
});

export class SwarmStorageMock extends SwarmStorage {
  constructor({ beeDebugEndpoint, beeEndpoint, chainId }: SwarmStorageConfig) {
    super({ beeDebugEndpoint, beeEndpoint, chainId });

    this._swarmClient = {
      bee: new Bee(beeEndpoint),
      beeDebug: beeDebugEndpoint ? new BeeDebug(beeDebugEndpoint) : undefined,
    };
  }

  get swarmClient(): SwarmClient {
    return this._swarmClient;
  }
}
