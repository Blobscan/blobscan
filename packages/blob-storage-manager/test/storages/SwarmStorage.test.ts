import { beforeAll, describe, expect, it } from "vitest";

import { testValidError } from "@blobscan/test";

import { env } from "../../src";
import { SwarmStorageMock as SwarmStorage } from "../../src/__mocks__/SwarmStorage";
import { BlobStorageError } from "../../src/errors";
import { BLOB_DATA, BLOB_HASH, SWARM_REFERENCE } from "../fixtures";

if (!env.BEE_ENDPOINT) {
  throw new Error("BEE_ENDPOINT test env var is not set");
}

const BEE_ENDPOINT = env.BEE_ENDPOINT;

describe("SwarmStorage", () => {
  let storage: SwarmStorage;

  beforeAll(() => {
    storage = new SwarmStorage({
      chainId: env.CHAIN_ID,
      beeEndpoint: BEE_ENDPOINT,
      beeDebugEndpoint: env.BEE_DEBUG_ENDPOINT,
    });
  });

  describe("constructor", () => {
    it("should create a new instance with the provided configuration", () => {
      expect(storage).toBeDefined();
      expect(storage.swarmClient.bee).toBeDefined();
      expect(storage.swarmClient.bee.url).toBe(env.BEE_ENDPOINT);
      expect(storage.swarmClient.beeDebug).toBeDefined();
      expect(storage.swarmClient.beeDebug?.url).toBe(env.BEE_DEBUG_ENDPOINT);
    });

    it("should not create beeDebug when beeDebugEndpoint is not set", () => {
      const newStorage = new SwarmStorage({
        chainId: env.CHAIN_ID,
        beeEndpoint: BEE_ENDPOINT,
      });

      expect(newStorage.swarmClient.beeDebug).toBeUndefined();
    });
  });

  describe("healthcheck", () => {
    it("should resolve successfully", async () => {
      await expect(storage.healthCheck()).resolves.not.toThrow();
    });

    testValidError(
      "should throw error if bee is not healthy",
      async () => {
        await storage.healthCheck();
      },
      BlobStorageError,
      {
        checkCause: true,
      }
    );

    testValidError(
      "should throw error if bee debug is not healthy",
      async () => {
        await storage.healthCheck();
      },
      BlobStorageError,
      {
        checkCause: true,
      }
    );
  });

  describe("getBlob", () => {
    it("should return the contents of the blob", async () => {
      const blob = await storage.getBlob(SWARM_REFERENCE);

      expect(blob).toEqual(BLOB_DATA);
    });
  });

  describe("storeBlob", () => {
    it("should store the blob in the bucket", async () => {
      const uploadReference = await storage.storeBlob(BLOB_HASH, BLOB_DATA);

      expect(uploadReference).toEqual(SWARM_REFERENCE);
    });

    testValidError(
      "should throw an error if no postage batches are available",
      async () => {
        await storage.storeBlob(BLOB_HASH, BLOB_DATA);
      },
      BlobStorageError,
      {
        checkCause: true,
      }
    );

    testValidError(
      "should throw an error if the bee debug endpoint is not available",
      async () => {
        const newStorage = new SwarmStorage({
          chainId: env.CHAIN_ID,
          beeEndpoint: BEE_ENDPOINT,
        });

        await newStorage.storeBlob(BLOB_HASH, BLOB_DATA);
      },
      BlobStorageError,
      {
        checkCause: true,
      }
    );
  });

  describe("tryGetConfigFromEnv", () => {
    it("should return a config object correctly", () => {
      const config = SwarmStorage.getConfigFromEnv({
        CHAIN_ID: env.CHAIN_ID,
        BEE_ENDPOINT: env.BEE_ENDPOINT,
        BEE_DEBUG_ENDPOINT: env.BEE_DEBUG_ENDPOINT,
      });

      expect(config).toEqual({
        chainId: env.CHAIN_ID,
        beeDebugEndpoint: env.BEE_DEBUG_ENDPOINT,
        beeEndpoint: env.BEE_ENDPOINT,
      });
    });

    testValidError(
      "should throw an error when a bee endpoint is not provided",
      () => {
        SwarmStorage.getConfigFromEnv({ CHAIN_ID: env.CHAIN_ID });
      },
      Error
    );
  });
});
