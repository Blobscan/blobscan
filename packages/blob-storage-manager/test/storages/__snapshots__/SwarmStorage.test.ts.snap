// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SwarmStorage > healthcheck > should throw error if bee debug is not healthy 1`] = `"SwarmStorageMock failed: Storage is not reachable"`;

exports[`SwarmStorage > healthcheck > should throw error if bee debug is not healthy 2`] = `[Error: Bee debug is not healthy: not ok]`;

exports[`SwarmStorage > healthcheck > should throw error if bee is not healthy 1`] = `"SwarmStorageMock failed: Storage is not reachable"`;

exports[`SwarmStorage > healthcheck > should throw error if bee is not healthy 2`] = `[Error: Bee is not healthy: not ok]`;

exports[`SwarmStorage > storeBlob > should throw an error if no postage batches are available 1`] = `"SwarmStorageMock failed: Failed to store blob with hash \\"0x0100eac880c712dba4346c88ab564fa1b79024106f78f732cca49d8a68e4c174\\""`;

exports[`SwarmStorage > storeBlob > should throw an error if no postage batches are available 2`] = `[Error: No postage batches available.]`;

exports[`SwarmStorage > storeBlob > should throw an error if the bee debug endpoint is not available 1`] = `"SwarmStorageMock failed: Failed to store blob with hash \\"0x0100eac880c712dba4346c88ab564fa1b79024106f78f732cca49d8a68e4c174\\""`;

exports[`SwarmStorage > storeBlob > should throw an error if the bee debug endpoint is not available 2`] = `[Error: Bee debug endpoint required to get postage batches.]`;

exports[`SwarmStorage > tryGetConfigFromEnv > should throw an error when a bee endpoint is not provided 1`] = `"SwarmStorageMock failed: No endpoint provided. You need to define BEE_ENDPOINT variable in order to use the Swarm storage"`;
